iceoryx_utils/include/iceoryx_utils/cxx/timer.hpp

```diff
@@ -0,0 +1,115 @@
+// Copyright (c) 2021 by Robert Bosch GmbH. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#ifndef IOX_UTILS_PERIODIC_TIMER_HPP
+#define IOX_UTILS_PERIODIC_TIMER_HPP
+
+#include "iceoryx_utils/cxx/expected.hpp"
+#include "iceoryx_utils/internal/units/duration.hpp"
+#include "iceoryx_utils/posix_wrapper/semaphore.hpp"
+#include <thread>
+
+using namespace iox::units;
+using namespace iox::units::duration_literals;
+
+namespace iox
+{
+namespace cxx
+{
+/// @brief This enum class offers the timer state events.
+/// If the timer is active the TimerEvent has ENABLED as the state.
+/// If tht timer is inactive the TimerEvent has DIABLED as the state.
+enum class TimerState : uint8_t
+{
+    DISABLED,
+    ENABLED
+};
+/// @brief This enum class offers the timer state events.
+/// After the wait if the timer is disabled it returns STOP.
+/// After the wait if the timer is enabled and to be activated without delay it returns TICK.
+/// After the wait if the timer is enabled and to be activated with delay it returns TICK_DELAY.
+/// After the wait if the timer is enabled and to be activated with delay breaching the delay threshold it returns
+/// TICK_THRESHOLD_DELAY.
+enum class TimerEvent : uint8_t
+{
+    STOP,
+    TICK,
+    TICK_DELAY,
+    TICK_THRESHOLD_DELAY
+};
```

Comment:
I'm not sure this is a good approach. It makes the Timer cumbersome to use. The `TICK_THRESHOLD_DELAY` is potentially an error and should be in the error path of the expected.

I would stick to just `STOP` and `TICK` and keep this simple. For the `TICK_DELAY` and `TICK_THRESHOLD_DELAY` I would recommend to make that part of the `CatchupPolicy` like
```cpp
enum class CatchupPolicy
{
    IMMEDIATE_TICK,
    SKIP_TO_NEXT_TICK,
    ERROR
};
```
and 
```cpp
enum class ErrorCause{
    TICK_EXCEEDED_TIME_LIMIT,
    OTHER_ERRORS
};

struct TimerError {
    ErrorCause cause;
    Duration tickDelay;
};
```

This could be either be a ctor parameter or a parameter to the `wait` method.
```cpp
cxx::expected<TimerEvent, TimerError> wait(CatchupPolicy p) noexcept;
```
with the following usage
```cpp
Timer t{100_ms};
t.start();
t.wait(CatchupPolicy::ERROR).or_else([] (auto& err) {
    if (err.cause == ErrorCause::TICK_EXCEEDED_TIME_LIMIT && err.tickDelay > 5_ms) {...}
});
```

Maybe it's also a good idea to completely get rid of the `TimerEvent` and handle the `STOP` case also with the `TimerError`, which would need to be renamed in this case.

Reply:
well, on the other hand a
```cpp
bool keepRunning = true;
while (keepRunning) {
    t.wait(CatchupPolicy::IMMEDIATE_TICK).and_then([&](TimerEvent e) {
        if(e == TimerEvent::STOP) {
            keepRunning = false;
            return;
        }
        // do stuff
    });
}
```
is also okay to use

Reply:
@elBoberido Actually this concept was proposed by @bishibashiB. Bernhard could you pls take a look the proposal by Bob and suggest?

Reply:
Nice @elBoberido 😁 for bringing in a `policy` config 'handshake'. 
First I understand that using an (optional) threshold to differ 'critical' from 'less-critical' delays should be continued?!
Then regarding the policy / error handling IMO you have the perfect overview to bed it into iceoryx codebase: ... as long as the user can read the timer 'return' ...   
Policy via wait()-parameter will hold more freedom to the user (than c'tor) I guess.
@shankar-in you also want to add something to it?

Reply:
Yes. I do agree that we don't introduce more complexity rather provide easy API for configuration of the timers. 

I understand that we pass the `policy` as a parameter to `wait()` as suggested by @bishibashiB. This also allows to modify the policy from the user side whenever needed. 

So `IMMEDIATE_TICK` precisely tries to call the next activation time and if the execution takes it returns error. The error returned will also have the duration of time of delay.

The `SKIP_TO_NEXT_TICK` in case of delay will get activated in the next available slot until then it blocks.

@elBoberido  Is my understanding right?

Reply:
@bishibashiB this is just loud thinking since it was similar in the old `Timer`. Maybe it would have been a good idea to create a design document before implementing it, then we could have discussed the API and feature scope in more detail. Maybe it's still a good idea to create a design document alongside this PR.

@shankar-in not quite. The idea is to have the non-error path of the `expected` handle only the happy path and maybe the stop path. The happy path would be when there is no delay or if you don't care about the delay. `IMMEDIATE_TICK` and `SKIP_TO_NEXT_TICK` would be such cases. The former would mean to immediately return with a `TICK` if the target time is already missed and the latter to skip the current tick and wait for the next scheduled one. `CatchupPolicy::ERROR` would mean an immediate return from `wait` if the target time is missed, but in this case in the error path of the `expected`. Instead of just returning an error, the delay from the target time would also be returned and it's up to the user if that is still acceptable. In the future, ff we notice that the use case with the delay is used quite often, we could introduce something like a `RelaxedTimer`/`FuzzyTimer` which would be just a small wrapper around the `Timer` from this PR and make the handling more convenient. Basically just hiding the `if (err.cause == ErrorCause::TICK_EXCEEDED_TIME_LIMIT && err.tickDelay > 5_ms) {...}` in a method which would call `Timer::wait(...)` beforehand.

Reply:
since a picture is more worth than thousand words, some ASCII art :)
```
Scheduled Ticks               |          |          |          |          |          |
IMMEDIATE_TICK Execution      [-------]  [----]     [-------------][-------][-----]
SKIP_TO_NEXT_TICK Execution   [-------]  [----]     [-------------]       [-------]  [-----]
ERROR Execution               [-------]  [----]     [-------------] <- return with error and let the user decide if the delay is acceptable
```

Reply:
So @bishibashiB I will be reworking on the `delay` and this will be handled as @elBoberido suggested with the `CatchupPolicy`

Reply:
@elBoberido I have a clarification in SKIP_TO_NEXT policy. I guess we also need to handle if the execution time took very long than the interval set.

+--------------------------+--------------+--------------+
|                          | Case 1       | Case 2       |
+--------------------------+--------------+--------------+
| Initial Time             | 10           | 10           |
+--------------------------+--------------+--------------+
| Catch-up Policy          | SKIP_TO_NEXT | SKIP_TO_NEXT |
+--------------------------+--------------+--------------+
| Interval                 | 50           | 50           |
+--------------------------+--------------+--------------+
|                          |              |              |
+--------------------------+--------------+--------------+
| 1st Cycle Execution Time | 70           | 150          |
+--------------------------+--------------+--------------+
| Expected Activation      | 60           | 60           |
+--------------------------+--------------+--------------+
| Time After Execution     | 80           | 160          |
+--------------------------+--------------+--------------+
| Difference for Activation | 20           | 100          |
+--------------------------+--------------+--------------+
| Wait()                   | 30           | 0            |
+--------------------------+--------------+--------------+
| Next Activation          | 110          | 160          |
+--------------------------+--------------+--------------+

Reply:
@shankar-in could you please wrap your ASCII art in markdown code blocks to get a monospace font?

Reply:
`
|                          	| Case 1       	| Case 2       	|
|--------------------------	|--------------	|--------------	|
| Initial Time             	| 10           	| 10           	|
| Catch-up Policy          	| SKIP_TO_NEXT 	| SKIP_TO_NEXT 	|
| Interval                 	| 50           	| 50           	|
|                          	|              	|              	|
| 1st Cycle Execution Time 	| 70           	| 150          	|
| Expected Activation      	| 60           	| 60           	|
| Time After Execution     	| 80           	| 160          	|
| Difference of Activation 	| 20           	| 100          	|
| Wait()                   	| 30           	| 0            	|
| Next Activation          	| 110          	| 160          	|`

@elBoberido Thanks for the hint.

Reply:
@shankar-in 
I guess you would have something like this
```
Scheduled Ticks               |     a    |     b    |     c    |     d    |     e    |
IMMEDIATE_TICK Execution      [---a---]  [------------b-----------][-c-][-d--][-e--] [---]
SKIP_TO_NEXT_TICK Execution   [---a---]  [------------b-----------]       [---e---]  [-----]
ERROR Execution               [---a---]  [------------b-----------] <- return with error and let the user decide if the delay is acceptable
```

Reply:
@elBoberido  Yes. I am doing in that. Thank you.